numba_rvsdg/mock_asm.py

```diff
@@ -0,0 +1,304 @@
+    Therefore, a program behavior is known statically.
+
+    """
+
+    def __init__(self, outbuf: IO[str]):
+        self._outbuf = outbuf
+        self._last_ctr = None
+        self._ctrmap: dict[int, int] = {}
+
+    def run(self, instlist: list[Inst], max_step: int | None = None) -> bool:
+        """Execute a list of instructions.
+
+        Each VM instance should only call `.run()` once.
+
+        Program starts at PC=0 and runs until the end of the instruction list
+        (PC >= length of instruction list) or `max_step` is reached
+
+        Returns whether program terminates by reaching the end.
+        """
+        step = 0
+        pc = 0
+        while pc < len(instlist):
+            if max_step is not None and step >= max_step:
+                return False
+            pc = self.eval_inst(pc, instlist[pc])
+            step += 1
+
+        return True
+
+    def eval_inst(self, pc: int, inst: Inst) -> int:
+        """Evaluate a single instruction.
+
+        Returns the next PC.
+        """
+        outbuf = self._outbuf
+        ctrmap = self._ctrmap
+        if inst.opcode == Opcode.PRINT:
+            print(inst.operands.text, file=outbuf)
+            return pc + 1
+        elif inst.opcode == Opcode.GOTO:
+            pc = inst.operands.jump_target
+            return pc
+        elif inst.opcode == Opcode.CTR:
+            operands: CtrOperands = inst.operands
+            # Set default state to the counter if not already defined
+            ctr = ctrmap.setdefault(pc, operands.counter)
+            # Decrement counter
+            ctrmap[pc] = ctr = max(0, ctr - 1)
+            # Store to LAST_CTR
+            self._last_ctr = ctr
+            return pc + 1
+        elif inst.opcode == Opcode.BRCTR:
+            operands: BrCtrOperands = inst.operands
+            # Read the LAST_CTR
+            ctr = self._last_ctr
+            assert ctr >= 0
+            # Do the jump
+            if ctr != 0:
+                return inst.operands.true_target
+            else:
+                return inst.operands.false_target
+        else:
+            raise AssertionError(f"invalid instruction {inst}")
+
+
+def test_mock_asm():
+    asm = """
+    print Start
+    goto A
+label A
+    print A
+    ctr 10
+    brctr A B
+label B
+    print B
+    """
+
+    instlist = parse(asm)
+    assert instlist[0].operands.text == "Start"
+    assert instlist[1].operands.jump_target == 2
+    assert instlist[2].operands.text == "A"
+    assert instlist[3].operands.counter == 10
+    assert instlist[4].operands.true_target == 2
+    assert instlist[4].operands.false_target == 5
+    assert instlist[5].operands.text == "B"
+
+    with StringIO() as buf:
+        VM(buf).run(instlist)
+        got = buf.getvalue().split()
+
+    expected = ["Start", *(["A"] * 10), "B"]
+    assert got == expected
+
+
+def test_double_exchange_loop():
+    asm = """
+    print Start
```

Comment:
I would recommend using the `textwrap.dedent` function here such that the `asm` is indented like the Python code that defines it.

Reply:
indentation shouldn't matter. i probably could have indent the asm to the same level to the python source

Reply:
done